contracts/vortex/CarbonVortex.sol

```diff
@@ -252,37 +858,109 @@ contract CarbonVortex is ICarbonVortex, Upgradeable, ReentrancyGuardUpgradeable,
+        // check if the maximum amount of halflifes have been reached
+        if (timeElapsed / currentPriceDecayHalfLife >= 128) {
+            return true;
+        } else {
+            return false;
         }
     }
 
-    function _validateTokens(Token[] calldata tokens) private view {
+    /**
+     * @dev returns the token amount available for trading
+     */
+    function _amountAvailableForTrading(Token token) private view returns (uint128) {
+        if (token == _targetToken) {
+            return _targetTokenSaleAmount.current;
+        } else {
+            return uint128(token.balanceOf(address(this)));
+        }
+    }
+
+    /**
+     * @dev validate token helper
+     */
+    function _validToken(Token token) private view {
+        // validate trading is enabled for token
+        if (!_tradingEnabled(token)) {
+            revert TradingDisabled();
+        }
+        // validate pair isn't disabled
+        if (_disabledPairs[token]) {
+            revert PairDisabled();
+        }
+    }
+
+    function _validateTokens(Token[] calldata tokens) private pure {
         uint256 len = tokens.length;
         if (len == 0) {
             revert InvalidTokenLength();
         }
         for (uint256 i = 0; i < len; i = uncheckedInc(i)) {
             Token token = tokens[i];
+            // revert for invalid token address
+            if (token == Token.wrap(address(0))) {
+                revert InvalidToken();
+            }
             // validate token has no duplicates
             for (uint256 j = uncheckedInc(i); j < len; j = uncheckedInc(j)) {
                 if (token == tokens[j]) {
                     revert DuplicateToken();
                 }
             }
-            // validate token can be traded on V3
-            if (token != _bnt && _bancorNetwork.collectionByPool(token) == address(0)) {
-                revert InvalidToken();
-            }
+        }
+    }
+
+    /**
+     * @dev validate token helper
+     */
+    function _validPrice(Price memory price) private pure {
+        if (price.sourceAmount == 0 || price.targetAmount == 0) {
+            revert InvalidPrice();
+        }
+    }
+
+    /**
+     * @dev return true if trading is enabled for token
+     */
+    function _tradingEnabled(Token token) private view returns (bool) {
+        return _tradingStartTimes[token] != 0;
+    }
+
+    /**
+     * @dev allocates the rewards to caller
+     */
+    function _allocateRewards(address sender, Token[] memory tokens, uint256[] memory rewardAmounts) private {
+        uint256 len = tokens.length;
+        for (uint256 i = 0; i < len; i = uncheckedInc(i)) {
+            Token token = tokens[i];
+            uint256 rewardAmount = rewardAmounts[i];
+            // transfer the rewards to caller
+            // safe due to nonReentrant modifier (forwards all available gas in case of ETH)
+            token.unsafeTransfer(sender, rewardAmount);
         }
```

Comment:
This function is only called from function `execute`, so it might be worth implementing it directly inside that function. Perhaps even as part of the loop over tokens in there (provided that we can indeed allocate rewards on the fly and not only at the end of that loop).
It would also make it easier to understand the comment about being safe due to `nonReentrant` modifier.

Reply:
easier to read if it's extracted in a separate function, 
also best to allocate rewards at the end of all state changes (standard security pattern)